### PR TITLE
50 show cmd output in floating window

### PR DESCRIPTION
**Pull request by ben-hansske**
The code does what it should. You can run a execute a shell command and it will output to a terminal buffer inside a floating window. It can also run any other lua function that outputs a string via `vim.api.nvim_chan_send`. The file `dashboard-term.lua` gives a short example on how to configure it and how to write a function that outputs to the terminal buffer. However there have been a couple of issues that I have not been able to solve or would like to hear your feedback:

- I am using `plenary.job` to run the external command. Is it ok to pull in an external dependency? Should I make it an optional dependency?
- I had to specify the command to run with command and the arguments in a list. As this would not allow for piping, I thought the best way to solve this would be to run `sh -c` and then also pass the command as a string, so that it will be executed by the shell and can contain pipes.
- The position of the floating window gets out of sync, when you scroll with <C-e> or <C-y>
- If either `:vsplit` or `:split` is executed, no second floating window will be spawned.
- In line 264 of `alpha.lua` I had to set `modifiable` on the alpha buffer to true, because I got error messages otherwise. I do not really understand why I had to do this, because I thought with the functions called I would not even touch it? Just felt weird to me, that's why I tell you here. I am not very experienced with lua and neovims api.

I am using this for almost a week now and did not encounter further problems. Please let me know what you think about this.

### CODE REVIEW

**Comment by goolord** [comment]
thank you so much for writing all of this ! i'm gonna go ahead and split the terminal code into its own module, also i renamed the element type to "term"

i'll try testing this and pushing whatever fixes i can later, too

**Comment by goolord** [comment]
printing out is also awfully slow on my machine for some reason, this is something else to investigate

File: lua/alpha/themes/dashboard-term.lua [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            end
+        end
+        output()
+    end
+end
+
+local command_header = {
+    type = "term",
+
+    --on_channel_opened = alpha.terminal_fillers.shell_command("echo - alpha.nvim - | figlet | lolcat -f"),
```
Comment (ben-hansske) [inline lua/alpha/themes/dashboard-term.lua:34]: Sorry, I probably have not explained myself well enough. `on_channel_openend` just gets a function passed that eventually fills the buffer. `animated_text_writer` was kind of a demo on how to use a lua function to draw
animated colored text. It uses `vim.defer_fn` to output just a single character every 10ms. The function
`alpha.terminal_fillers.raw_string` does the same thing, except it outputs the text immediately. If a shell command is wanted, `alpha.terminal_fillers.shell_command` would be the way to go.

Comment (goolord) [reply lua/alpha/themes/dashboard-term.lua:34]: oooooooohhh cool

**Comment by goolord** [comment]
another todo: hide the cursor in the terminal buffer

**Comment by ben-hansske** [comment]
There is one more issue I forgot to mention up there: `lolcat` for example sees that it does not output to a terminal and does not print colors unless it is forced with `-f`. I already checked the `$TERM` variable, but it looked fine. I think the best way to get around this might be to use `vim.fn.termopen` instead of `vim.api.nvim_open_term`. I'll give it a try.

**Comment by danielnehrig** [comment]
let me know if you need/want testing from my side

**Comment by ben-hansske** [comment]
Sorry for not responding in a while! Testing is of course much appreciated.

With the last commit I was able to sort out the following issues:
- 'plenary.job' is no longer required
- There is no longer the need to use 'sh -c' to interpret the command, the default shell is now used.
- The executed programs now know that there output goes to a pseudo terminal and therefore handle colors correctly.
- If a program does not exit, it will be terminated when the auxiliary window is closed

Remaining issues are still:
- Scrolling inside the alpha buffer does not effect the position of the auxiliary window. I tried to use 'vim.fn.line' to get the current scroll position but this function does not work inside the alpha buffer for some reason
- The window is not cloned when doing a split inside alpha. Implementing this could get complicated.

I am not sure how much you care about these two remaining issues. I would leave them as they are if you don't mind?

**Comment by danielnehrig** [comment]
any update here what is missing ?

**Comment by goolord** [comment]
i'll try messing with animating the window position, but if it's not fruitful i'll just merge it, i think

**Comment by goolord** [comment]
`line` does work, i don't really understand why `line('.')` always returns 0 (even in normal files) but `line('w0')` is more like what we want i think

File: lua/alpha/term.lua [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        -- I have no clue why I need to do this, but otherwise it gives errors :/
+        vim.api.nvim_buf_set_option(state.buffer, "modifiable", true)
+
+        state.aux_windows[itemid] = window
+    elseif type(state.aux_windows[itemid]) == "table" then
+        local window = state.aux_windows[itemid]
+        vim.api.nvim_win_set_config(window.win, win_options)
+    end
+
```
Comment (ben-hansske) [inline lua/alpha/term.lua:126]: I tried adding
```suggestion
	print(state.window, vim.fn.line('w0', state.window))
```
Now matter how the current scroll position is, i get the same output:
```
1000 1
```
Am I doing something wrong here or why does this not work? Manually calling `line` on the other hand gives the correct scroll values.

**Comment by olimorris** [comment]
Hey @ben-hansske just coming to this now. Hadn't realised we had this awesome pull request in the works and more than happy to test this out. Is there anything we can borrow from [dashboard.nvim's](https://github.com/glepnir/dashboard-nvim/blob/master/lua/dashboard/preview.lua#L43) recent refactor?

**Comment by ben-hansske** [comment]
Hi @olimorris, I am quite busy right now, but I will try to schedule some time the next week looking into the recent developement on dashboard.nvim.

**Comment by ben-hansske** [comment]
There is an example theme: https://github.com/ben-hansske/alpha-nvim/blob/50-show-cmd-output-in-floating-window/lua/alpha/themes/dashboard-term.lua
To use it
```
require'alpha'.setup(require'alpha.themes.dashboard-term'.opts)
```
should be enough.

Feel free to play around with it! The lines 37 - 56 are probably what you are looking for.

**Comment by goolord** [comment]
closed in favor of https://github.com/goolord/alpha-nvim/pull/124 for the moment

**Pull request closed without merging**